cmd/disk-cache-backend.go

```diff
@@ -396,6 +425,39 @@ func (c *diskCache) bitrotWriteToCache(ctx context.Context, cachePath string, re
 	return bytesWritten, nil
 }
 
+func newCacheEncryptReader(content io.Reader, bucket, object string, metadata map[string]string) (r io.Reader, err error) {
+	objectEncryptionKey, err := newCacheEncryptMetadata(bucket, object, metadata)
+	if err != nil {
+		return nil, err
+	}
+
+	reader, err := sio.EncryptReader(content, sio.Config{Key: objectEncryptionKey[:], MinVersion: sio.Version20})
+	if err != nil {
+		return nil, crypto.ErrInvalidCustomerKey
+	}
+	return reader, nil
+}
+func newCacheEncryptMetadata(bucket, object string, metadata map[string]string) ([]byte, error) {
```

Comment:
That function can just call `newEncryptMetadata` and then add the `metadata[SSECacheEncrypted] = ""` (if necessary at all) - or am I missing something?

Reply:
@aead, newEncryptMetadata uses the globalKMS whereas cache needs to use its own KMS, hence the need to have a separate function. Otherwise cache logic will spill over to handler and encryption-v1.go